### DISCUSSION TITLE
Managing images

### DISCUSSION THREAD

**Discussion by llermaly** [hagopj13/node-express-boilerplate#54] [Q&A]
Hey, what's you approach to manage images in terms of data flow? Let's say we are uploading a user avatar using S3 and storing the URL in the database.

Or we want to set images to the products. 

I found many information about uploading the images as the only thing but not the best practice to upload it as a part of a document.

Most of people mentioned a new route like /product/:id/upload-image , but then where do we put the code and how we handle the cases when S3 fails? We should wait until the product is created to add the image and then confirm? , just upload the product without image?

I found this gist but I'm not sure on how to handle the data flow the proper way :
https://gist.github.com/SylarRuby/b60eea29c1682519e422476cc5357b60

I would appreciate some guidance on whats your vision on this.

Thanks!

**Comment by hagopj13** [comment]
Hey @llermaly, uploading the image to an S3 bucket and storing the url of the image as part of the product info sounds like a good idea. You can also use something like `multer` to store the image as a file on your server.

Concerning whether to do it in one shot or to create separate API's, then that would depend on the process of creating products. If you need to reuse the image uploading functionality for other objects, then you can make it a separate API. Otherwise, if it will only be used for creating products, then you can do it in one API. `multer` would come in really handy for the latter. After a quick search on Google, I saw this: https://medium.com/@nitinpatel_20236/image-upload-via-nodejs-server-3fe7d3faa642 . I hope it can give you some hints.
